src/workerd/api/node/buffer.c++

```diff
@@ -0,0 +1,350 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
+
+#include "buffer.h"
+#include <workerd/jsg/buffersource.h>
+#include <workerd/api/crypto-impl.h>
+#include <kj/encoding.h>
+#include <algorithm>
+
+// These are defined by <sys/byteorder.h> or <netinet/in.h> on some systems.
+// To avoid warnings, undefine them before redefining them.
+#ifdef BSWAP_2
+# undef BSWAP_2
+#endif
+#ifdef BSWAP_4
+# undef BSWAP_4
+#endif
+#ifdef BSWAP_8
+# undef BSWAP_8
+#endif
+
+#if defined(_MSC_VER)
+#include <intrin.h>
+#define BSWAP_2(x) _byteswap_ushort(x)
+#define BSWAP_4(x) _byteswap_ulong(x)
+#define BSWAP_8(x) _byteswap_uint64(x)
+#else
+#define BSWAP_2(x) ((x) << 8) | ((x) >> 8)
+#define BSWAP_4(x)                                                            \
+  (((x) & 0xFF) << 24)  |                                                     \
+  (((x) & 0xFF00) << 8) |                                                     \
+  (((x) >> 8) & 0xFF00) |                                                     \
+  (((x) >> 24) & 0xFF)
+#define BSWAP_8(x)                                                            \
+  (((x) & 0xFF00000000000000ull) >> 56) |                                     \
+  (((x) & 0x00FF000000000000ull) >> 40) |                                     \
+  (((x) & 0x0000FF0000000000ull) >> 24) |                                     \
+  (((x) & 0x000000FF00000000ull) >> 8)  |                                     \
+  (((x) & 0x00000000FF000000ull) << 8)  |                                     \
+  (((x) & 0x0000000000FF0000ull) << 24) |                                     \
+  (((x) & 0x000000000000FF00ull) << 40) |                                     \
+  (((x) & 0x00000000000000FFull) << 56)
+#endif
+
+namespace workerd::api::node {
+
+namespace {
+template <typename T>
+void SwapBytes(kj::ArrayPtr<kj::byte> bytes) {
+  uint32_t len = bytes.size() / sizeof(T);
+  T* data = reinterpret_cast<T*>(bytes.begin());
+  for (uint32_t i = 0; i < len; i++) {
+    if constexpr (kj::isSameType<T, uint16_t>()) {
+      data[i] = BSWAP_2(data[i]);
+    } else if constexpr (kj::isSameType<T, uint32_t>()) {
+      data[i] = BSWAP_4(data[i]);
+    } else {
+      data[i] = BSWAP_8(data[i]);
+    }
+  }
+}
+
+enum class Encoding {
+  ASCII,
+  LATIN1,
+  UTF8,
+  UTF16LE,
+  BASE64,
+  BASE64URL,
+  HEX,
+};
+
+Encoding getEncoding(kj::StringPtr encoding) {
```

Comment:
For later, something we may want to consider is a jsg utility for automatically mapping webidl style enums.